### ISSUE TITLE
support for upcoming Typescript path mappings

### ISSUE THREAD

**Issue by opensrcken** [ivogabe/gulp-typescript#299]
I'm running against the nightly build of Typescript, and there is an incompatibility issue when using the newly added Typescript path mappings (https://github.com/Microsoft/TypeScript/issues/5039#pathMappings), specifically with commonBaseDiff in compiler.js. Leveraging path mappings seems to produce unexpected output for the following:
```
var expected = this.project.input.commonBasePath; 
var real = this.project.input.commonSourceDirectory;
```

This subsequently causes the following code in output.js’s #addOrMergeFile method to always evaluate to true, which means that the compiled output is not emitted.

`if (!this.project.compiler.correctSourceMap(file.sourceMap)) { … }`

I don’t have a firm enough grasp on the inner workings of this logic to confidently create a proper fix at the moment. Can someone with more domain expertise look into this?

For now, I’m commenting out the correctSourceMap check, and hoping it does not screw me over.

ALSO:

Renaming a typescript file while it’s being watched causes the following error:
```
Error: error TS6053: File ‘{{FILENAME}}’ not found.
    at DestroyableTransform.<anonymous> (.../node_modules/gulp-typescript/release/project.js:94:33)
    at emitNone (events.js:72:20)
    at DestroyableTransform.emit (events.js:166:7)
    at finishMaybe (.../node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:511:14)
    at endWritable (.../node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:521:3)
    at DestroyableTransform.Writable.end (.../node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:486:5)
    at DestroyableTransform.onend (.../node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:545:10)
    at DestroyableTransform.g (events.js:260:16)
    at emitNone (events.js:72:20)
    at DestroyableTransform.emit (events.js:166:7)
```

**Comment by sandeepgs3** [comment]
any updates on this?

**Comment by ivogabe** [comment]
@sandeepgs3 No, I haven't had time to work on this.

**Comment by alexeagle** [comment]
another related issue, I use the new `baseUrl` to make it possible to resolve imports from another location,
```
"baseUrl": "../../dist/js/cjs",
```
and even though I can run `$(npm bin)/tsc -p tools/compiler_cli` successfully, this gulp config fails
```
  var tsc = require('gulp-typescript');
  var tsProject = tsc.createProject('tools/compiler_cli/tsconfig.json', {
    typescript: require('typescript')
  });
  var tsResult = tsProject.src().pipe(tsc(tsProject));
  return tsResult.js.pipe(gulp.dest('dist/tools/compiler_cli'));
```

with
```
tools/compiler_cli/linker.ts(6,27): error TS2307: Cannot find module 'angular2/src/compiler/compiler'.
```

My guess is that `tsProject.src()` creates a gulp file stream, but it doesn't include files referenced outside the directory where tsconfig lives.

**Comment by ivogabe** [comment]
I've implemented this in fb4ffc0a72139db1a709dd36ebb9ef92543bdfa6. The path mapping settings needed to be converted to absolute paths. For the calculation of the common source directory, absolute paths were necessary (as @opensrcken pointed out). I plan to release a new version this weekend.
